vm/core/hash.go

```diff
@@ -14,14 +15,25 @@ func SigningBody(genesis, tx []byte) []byte {
 	return full
 }
 
-// ComputePrincipal address as the last 20 bytes from blake3(scale(template || args)).
-func ComputePrincipal(template Address, args scale.Encodable) Address {
+// ComputePrincipal address as the last 24 bytes of Hash(template || blob).
+// See https://github.com/spacemeshos/go-spacemesh/issues/6420 for more details.
+func ComputePrincipalFromBlob(template types.Address, blob []byte) Address {
 	hasher := hash.GetHasher()
 	defer hash.PutHasher(hasher)
-	encoder := scale.NewEncoder(hasher)
-	template.EncodeScale(encoder)
-	args.EncodeScale(encoder)
+	hasher.Write(template[:])
+	hasher.Write(blob)
 	sum := hasher.Sum(nil)
-	rst := types.GenerateAddress(sum[12:])
-	return rst
+	return types.GenerateAddress(sum)
+}
```

Comment:
I'm unsure if using `types.GenerateAddress` is OK here, see its implementation: https://github.com/spacemeshos/go-spacemesh/blob/40d4ba62fddfc2c81a19556b80400166da8bd39b/common/types/address.go#L132-L139

It does more than taking the upper 24B - it also reserves the lower 4B of the returned address (they are zeros)

Reply:
isn't it 20B (used) + 4B (reserved) = 24B total?

Reply:
yes

Reply:
then that should be fine, since Athena also uses 24 byte addresses. I don't understand your initial comment here.

Reply:
There are 2 problems with using `types.GenerateAddress()`:
1. It supposedly creates an address from a public key:
    ```go
    // GenerateAddress generates an address from a public key.
    func GenerateAddress(publicKey []byte) Address {
    ```
2. It's implementation doesn't match `HASH(template_address, immutable_state)` as it also "reserves" those 4B. 

IMHO what you need here is 
```go
var principal Address
copy(principal[:], sum[:24])
return principal
```

instead of 
```go
return types.GenerateAddress(sum)
```